src/main/java/hudson/plugins/ec2/ssh/EC2UnixLauncher.java

```diff
@@ -399,6 +412,69 @@ private boolean bootstrap(EC2Computer computer, TaskListener listener, SlaveTemp
+        int tries = 13; // 3 minutes before giving up
+        int sleepMillis = 15000;
+        boolean status = false;
+        String msg = "";
+        final String baseMsg = "Node " + computer.getName();
+        try {
+            while (tries-- > 0) {
+                switch (computer.getState()) {
+                case PENDING:
+                    msg = baseMsg + " is still pending/launching, waiting " + sleepMillis/1000 + "s";
+                    // Assume a pending agent will start and allow Jenkins to SSH into it
+                    logInfo(computer, listener, msg);
+                    status = true;
+                    break;
+                case STOPPING:
+                    msg = baseMsg + " is still stopping, waiting " + sleepMillis/1000 + "s";
+                    logInfo(computer, listener, msg);
+                    break;
+                case RUNNING:
+                    msg = baseMsg + " is ready";
+                    logFiner(computer, listener, msg);
+                    status = true;
+                    tries = 0;
+                    break;
+                case STOPPED:
+                    msg = baseMsg + " is stopped, sending start request";
+                    logInfo(computer, listener, msg);
+                    AmazonEC2 ec2 = computer.getCloud().connect();
+                    List<String> instances = new ArrayList<String>();
+                    instances.add(computer.getInstanceId());
+                    StartInstancesRequest siRequest = new StartInstancesRequest(instances);
+                    StartInstancesResult siResult = ec2.startInstances(siRequest);
+                    msg = baseMsg + ": sent start request, result: " + siResult;
+                    logInfo(computer, listener, msg);
+                    break;
+                case SHUTTING_DOWN:
+                    msg = baseMsg + " is still shutting down, waiting " + sleepMillis/1000 + "s";
+                    logInfo(computer, listener, msg);
+                    break;
+                case TERMINATED:
+                    // abort
+                    msg = baseMsg + " is terminated or terminating, aborting launch";
+                    logInfo(computer, listener, msg);
+                    status = false;
+                    tries = 0;
+                    break;
+                default:
+                    msg = baseMsg + " is in an unknown state, retrying in" + sleepMillis/1000 + "s";
+                    logInfo(computer, listener, msg);
+                }
+                if (tries != 0) {
+                    Thread.sleep(sleepMillis);
+                }
+            }
+        } catch (AmazonClientException e) {
+            e.printStackTrace(listener.error(e.getMessage()));
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            e.printStackTrace(listener.error(e.getMessage()));
+        }
+        return status;
+    }
```

Comment:
I like that you created a separate method for this logic.  Very nice work @mkozell !